cupy/core/reduction.pxi

```diff
@@ -175,24 +175,97 @@ cpdef list _get_inout_args(
+    params = _get_kernel_params(params, args_info)
 
-    t = (_get_typename(in_arg_dtype), _get_typename(out_arg_dtype))
-    type_preamble = 'typedef %s type_in0_raw; typedef %s type_out0_raw;' % t
+    type_preamble = '\n'.join(
+        'typedef %s %s;' % (_get_typename(v), k)
+        for k, v in types)
 
-    params = _get_kernel_params(params, args_info)
     return _get_simple_reduction_kernel(
         name, block_size, reduce_type, params, identity,
-        routine[0], routine[1], routine[2],
+        map_expr, reduce_expr, post_map_expr,
         type_preamble, input_expr, output_expr, _preamble, options)
 
 
-class simple_reduction_function(object):
+cdef class _AbstractReductionKernel:
+
+    cdef str name
+    cdef str identity
+
+    cpdef _call(
+            self, in_args, out_args, params,
+            a_shape, axis, dtype,
+            bint keepdims, bint reduce_dims,
+            stream):
+        cdef tuple reduce_axis, out_axis
+        cdef Py_ssize_t contiguous_size
+        cdef Py_ssize_t block_size, block_stride, out_block_num
+        cdef ndarray ret
+        cdef function.Function kern
+
+        if self.identity is None and 0 in a_shape:
+            raise ValueError(('zero-size array to reduction operation'
+                              ' %s which has no identity') % self.name)
+
+        if dtype is not None:
+            dtype = get_dtype(dtype).type
+
+        (
+            map_expr, reduce_expr, post_map_expr,
+            in_types, out_types, reduce_type,
+            types,
+        ) = self._get_expressions_and_types(in_args, out_args, dtype)
+
+        reduce_axis, out_axis = _get_axis(axis, len(a_shape))
+        out_shape = _get_out_shape(a_shape, reduce_axis, out_axis, keepdims)
+        out_args = self._get_out_args(out_args, out_types, out_shape)
+        ret = out_args[0]
+        if ret.size == 0:
+            return ret
+
+        in_args = [x if isinstance(x, ndarray) else
+                   _scalar.get_scalar_from_numpy(x, t)
+                   for x, t in zip(in_args, in_types)]
+        in_args, in_shape, contiguous_size = _get_permuted_args(
+            in_args, reduce_axis + out_axis, a_shape, None, len(out_axis))
+
+        in_indexer = Indexer(in_shape)
+        out_indexer = Indexer(out_shape)
+        block_size, block_stride, out_block_num = _get_block_specs(
+            in_indexer, out_indexer, contiguous_size)
+
+        inout_args = _get_inout_args(
+            in_args, out_args, in_indexer, out_indexer, block_stride,
+            params, reduce_dims)
+        args_info = _get_args_info(inout_args)
+
+        kern = self._get_kernel(
+            params, args_info, types,
+            map_expr, reduce_expr, post_map_expr, reduce_type,
+            block_size)
+        kern.linear_launch(
+            out_block_num * block_size, inout_args, 0, block_size, stream)
+        return ret
+
+    cpdef _get_expressions_and_types(self, in_args, out_args, dtype):
```

Comment:
Why not support them as `cdef` methods and write argument types explicitly?

Reply:
`cdef` methods cannot be overridden.
`dtype` argument does not have a specific type.

Reply:
Do you mean other arguments? 
I doubt typing tuples improves performance. I'll check.

Reply:
I only observe difference of 1-2 ns. I think it's negligible.